tools/tpm2_policyauthorize.c

```diff
@@ -0,0 +1,173 @@
+#include "tpm2_session.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+typedef struct tpm2_startauthsession_ctx tpm2_startauthsession_ctx;
+struct tpm2_startauthsession_ctx {
+   //File path for the session context data
+   const char *s;
+   //File path for the policy digest that will be authorized
+   const char *p;
+   //File path for the policy qualifier data
+   const char *q;
+   //File path for the verifying public key name
+   const char *u;
+   //File path for the verification ticket
+   const char *t;
+   //File path for storing the policy digest output
+   const char *o;
+};
+
+static tpm2_startauthsession_ctx ctx;
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'o':
+        ctx.o = value;
+        break;
+    case 'S':
+        ctx.s = value;
+        break;
+    case 'f':
+        ctx.p = value;
+        break;
+    case 'q':
+        ctx.q = value;
+        break;
+    case 'n':
+        ctx.u = value;
+        break;
+    case 't':
+        ctx.t = value;
+        break;
+    }
+    return true;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    static struct option topts[] = {
+        { "policy-file",       required_argument, NULL, 'o' },
+        { "session",           required_argument, NULL, 'S' },
+        { "input-policy-file", required_argument, NULL, 'f' },
+        { "qualifier",         required_argument, NULL, 'q' },
+        { "name",              required_argument, NULL, 'n' },
+        { "ticket",            required_argument, NULL, 't' },
+    };
+
+    *opts = tpm2_options_new("o:S:f:q:n:t:", ARRAY_LEN(topts), topts, on_option,
+                             NULL, 0);
+
+    return *opts != NULL;
+}
+
+bool is_check_input_options_ok(void) {
+
+    if (!ctx.s) {
+        LOG_ERR("Must specify -S session file.");
+        return false;
+    }
+
+    if (!ctx.p) {
+        LOG_ERR("Must specify -a p name.");
+        return false;
+    }
+
+    if (!ctx.u) {
+        LOG_ERR("Must specify -c u name.");
+        return false;
+    }
+
+    return true;
+}
+
+int tpm2_tool_onrun(TSS2_SYS_CONTEXT *sapi_context, tpm2_option_flags flags) {
+
+    UNUSED(flags);
+
+    if (!is_check_input_options_ok()) {
+        return -1;
+    }
+
+    int rc = 1;
+
+    tpm2_session *s = tpm2_session_restore(sapi_context, ctx.s);
+    if (!s) {
+        return rc;
+    }
+
+    bool result = tpm2_policy_build_policyauthorize(sapi_context, s, ctx.p,
+                        ctx.q, ctx.u, ctx.t);
+    if (!result) {
+        LOG_ERR("Could not build tpm authorized policy");
+        goto out;
+    }
+
+    TPM2B_DIGEST policy_digest = TPM2B_EMPTY_INIT;
+    result = tpm2_policy_get_digest(sapi_context, s, &policy_digest);
+    if (!result) {
+        LOG_ERR("Could not build tpm policy");
+        goto out;
+    }
+
+    tpm2_util_hexdump(policy_digest.buffer, policy_digest.size);
+
+    if (ctx.o) {
+        result = files_save_bytes_to_file(ctx.o, (UINT8 *) &policy_digest.buffer,
```

Comment:
I think you can get tid of this cast

Reply:
Oooh that cast was dangerous. You were passing a BYTE** there. the buffer field was already a pointer and you took the address of.